Revamped money cheats

wolfreak99 opened the pull request:
For a long while the Money page for the cheats window has never changed. It has 2 buttons, clear loan, and 5000+ cash. if you wanted a high amount of money, you had to repeatedly click. If you were running a server in multiplayer and wanted to switch to no money, you had to go to the scenario options, hit no money, save the park, and reload it, or tell others to log out and log back in. I have decided to take a go at updating the money page.

This pull request features a few new cheats, modes, along with the ability to add or set the money (even over multiplayer). 

-**No money**
This is pretty self explanatory, it is taken from the scenario options window, but set inside a cheat to allow multiplayer usage. 
-**Add / Set money**
This is the same idea as the Add 5000+ button, only gone is the single choice of a number, I have added a spinner that increments by 1000, along with it being able to have a custom amount typed in (with support for commas, and dots.. The conversion method might be a little overdone, but as it stands, it works and is flexible enough to type in 10,000,000.00 without needing to rely on, say, counting 0's in 1000000000)
-**Ignore money**
This will allow you to continue to build stuff, even if you venture into bankruptcy.
-**Freebuilding mode**
When this is enabled, you can build stuff without the money actually impacting your in-game funds. Think of it as building onto a park while still being able to watch it financially profit. Instead of the red and green money effects, you will see greyed out money effects. I chose to kept the floating money because while in multiplayer, it helps viewing anything subtly placed. Despite Freebuilding not effecting the games actual money, it will still update players money spent in multiplayer mode. multiplayer money spent and the money effects was discussed and chosen to keep via gitter chat and various servers.

http://i.imgur.com/tENO6sD.gif
This is a preview of the functions (with the exception of manually typing in the money)

While I usually keep my stuff as works in progress, this is a finished pull request as far as my own ideas go. I had explicit intentions with this: import the no-money function for multiplayer, ditch the 5000 limit for money, and update it with some extra features to make opening the cheats window a more better experience. Once the code is reviewed, I will squash all the commits to one.

mrtnptrs commented:
Works fine for me.

Phyxion commented:
Few things:
> -No money
This is pretty self explanatory, it is taken from the scenario options window, but set inside a cheat to allow multiplayer usage. 
Perhaps the scenario option should be synced in multiplayer rather than adding another option.

>  -Add / Set money
Dots and commas depend on the users locale. US is 10,000.00 but there are also countries with 10.000,00.

wolfreak99 commented:
I can look into the idea of switching around the decimal and comma. Would you know any currencies that contain such syntax? As far as the no money cheat, as nice as it may be to have it in a debug multiplayer, the multiplayer has debugging completely off. There's no debug functionality yet and I wouldn't know how to begin a scratch category for debugging features, plus it feels like it would need more than just one button available in multiplayer, where as the cheats are all already multiplayer compatible. Basically, being the one to start multiplayer functionality for debugging feels too overwhelming and too much for an amateur programmer like myself, and It feels like it wouldn't be worth it

duncanspumpkin commented on src/windows/cheats.c:
```diff
@@ -756,29 +865,55 @@ static void window_cheats_rides_mouseup(rct_window *w, int widgetIndex)
-				if (get_forced_park_rating() >= 0)
-					set_forced_park_rating(park_rating_spinner_value);
+	if (w->page == WINDOW_CHEATS_PAGE_MISC) {
+		if (widget_is_pressed(w, WIDX_INCREASE_PARK_RATING) || widget_is_pressed(w, WIDX_DECREASE_PARK_RATING))
```
This is not how you do the repeat button. There is a flag you enable to do all of this. Can't remember the name but I recently made the change in the rate input window so go check that.

IntelOrca replied:
`w->hold_down_widgets`

wolfreak99 commented:
Honestly I knew about that hold down but I got worried that it might not had been accepted if I changed to it since it would involve changing the core code. Is it okay to change that? I remember reading that we shouldn't make too much changes to the decompiled version at least until it's done, and if so mark that it was changed. Are we still following that rule? Would a comment "//the following code was updated to use hold_down_widgets" be sufficient enough? I also have a feeling I might just be looking into the projects rules and strictness too much again. (I hope I don't sound like I'm making fun of the strictness, I'm not. It's not you guys, it's me, I'm just not used to group projects yet. Git, group collaboration, trying to match standards and ethics, c, this is all very new to me)

wolfreak99 commented:
Long comment short: it's okay to change the code including the way it previously was to use hold down widgets? Should I add a comment stating it was updated? Will it cause any trouble with reverse engineering at this point?

wolfreak99 commented:
@duncanspumpkin  @IntelOrca  I probably should have waited until you re-reviewed the code before squashing, apologies. but I have fixed the code to now use hold_down_widgets. I have also added in the feature into the changelog.

duncanspumpkin commented:
I don't quite see the reason you would want freebuild mode or ignore money surely the no money mode is all that is required. It feels like it's over complicating a problem that already has a satisfactory solution. I'm not opposed to it but would like to know if this is something people actually want. My main worry is that since money is used for many different things and windows these new modes may be forgotten and cause new bugs.

wolfreak99 commented:
The main idea behind those two modes was if you still want to view finances... see if your park is doing successful while you work on a big ride, etc etc. 
Are you talking about new bugs in the future? or talking about bugs as the current code stands? 
I checked around on multiplayer and seen that people would like the option to have freebuild mode, namely because of the idea of watching players action history costs, like if someone comes in and builds a block of land up and down repeatedly to try to drain the money. 

Maybe I could remove the two cheats for now and then we can see if people want them, and if so, I can put them in a seperate request?

IntelOrca commented:
> I checked around on multiplayer and seen that people would like the option to have freebuild mode, namely because of the idea of watching players action history costs, like if someone comes in and builds a block of land up and down repeatedly to try to drain the money.

This is a bug in multiplayer, the amount spent for each player should increase even if money is disabled.

wolfreak99 commented:
@IntelOrca  It looks like the problem is that all the game command functions still check for no money enabled or not, and then returns either 0 or the amount needed based on whether no money is enabled.

rvgr commented:
A quick comment on dots and commas: I don't know how you managed this but they are localized with strings 5151 and 5152: STR_LOCALE_THOUSANDS_SEPARATOR for the thousands separator and STR_LOCALE_DECIMAL_POINT for decimal separators.

For example in French we use a space as thousands separator and a comma as decimal separator.

Nubbie commented:
Potential fix for #3952

wolfreak99 commented:
@rvgr Apologies that it took so long to respond. Thank you for pointing those out! That will definitely be useful! I'll see about plugging those in in a few moments actually.. but, In french are your currencies still in a "solid first change second" format? Like, in US currency, you have "$1,234.56" which is 1234 dollars and 56 cents. would yours still be where the cents (or your currency equivalent to U.S. cents) will be after the dollars, or would cents be wrote out first? i.e. "$56,1.234" or 56 cents and 1234 dollars.. Or would it still be in the same left-to-right format, only with different seperator symbols? i.e. "$1.234,56" still being 1234 dollars and 56 cents. Hopefully my question makes sense.

rvgr commented:
@wolfreak99 In French it uses the left to right format with different separators, for example $1,234.56 would be 1 234,56$. We put the currency symbol after the amount, however as far as I know this isn't localized, it is defined in the currency settings (I use a custom currency to put the symbol as a suffix).

Throughout the localization file there are two variables named {CURRENCY} and {CURRENCY2DP} that are used to display various amounts in the UI with the right localized format and set currency, however I don't know to what it's linked.

Gymnasiast commented:
In case handling the different decimal separators doesn't work out, another option would to let the user type in the bare amount, and provide a conversion in the description. Basically:
```
Enter an amount (£ 1234.50 = 12345):
_
```
The `£ 1234.50` will be replaced according to the selected currency and locale, like this for example:
(Dutch with guilders)

```
Voer een bedrag in (fl. 1234,50 = 12345):
_
```

And like this:
(English with Japanese Yen)
```
Enter an amount (¥ 123,450 = 12345):
_
```

You could then filter out anything that isn't a number, convert to an int and use it.

IntelOrca commented:
This will be relevant if we want to provide text input to all price fields.

marijnvdwerf commented:
That's what Microsoft Word does, and it's rather annoying when you use the numpad to type `29.7`, and it doesn't get interpreted as `29 + DECIMAL_SEPERATOR + 7`.

Patrik356b commented:
Like pretty much every other application then.

wolfreak99 commented:
I believe I've updated it to use the decimal and thousand separator. For right now, I set the part that converts it from the number to the text box to not include the thousands or currency symbol, only the decimal separator (and only if necessary).. but @intelorca thanks for saying there could be future use for the conversion, that actually helps alot, considering I started to believe "maybe I'm overdoing this, I need to keep it simple, I doubt anyone will really format it by thousands when typing anyway", but the idea of modifiable numbers throughout the game does sound pretty cool and gives me more reason to do improve it.

One question though, is there any way to grab the currency symbol of the current currency? Like $ or ¥, that way I can filter out the symbol in the string to number conversion (much like the thousands separator gets filtered, the only technical separator thats needed for conversion is the decimal)

duncanspumpkin commented:
@wolfreak99 just filter out anything that isn't the decimal seperator or number. Then do the conversion. No need to know what the currency symbol is. I.e. `$1,234.12` would be read as `1234.12` and `4abc123.4abc` would be read as `4123.4`.

wolfreak99 commented:
How would be the best way to do that, @duncanspumpkin ?

duncanspumpkin commented:
A naive approach would be to do the following

```
text_ptr = text
text_len = text.length
while(char ch = *text_ptr++ != 0){
  text_len--;
  if (ch == DECIMAL_SEPERATOR ||
      (ch >= '0' && ch <= '9'))
    continue;

  memmov(text_ptr - 1, text_ptr, text_len);
  text_ptr[text_len] = 0;
}
``

I think that would work but I'm not completely sure it might need some modification to that memmove.

And then when you have it in that format you can split the text at the decimal seperator and read it as two variables. Would also need to add something for looking for a negative i guess.

wolfreak99 commented:
That looks like it would do the trick. Now out of curiosity, what file would be a good file to place that function in? I don't want to have it specifically in just the cheats window, especially if (as IntelOrca said) there is potential for the idea to be implemented in many fields. I've always wondered about where to place functions that could be broadly used. Would I be able to place it in util? Or what do you think?

Gymnasiast commented:
I think `src/input.c` would be the best place.

wolfreak99 commented:
for a second i thought you were saying place a small function that checks for numbers and decimal inside input, but then i realized he didn't comment it as a dedicated function but more of an internal code to put inside the function.

Gymnasiast commented:
Well, splitting it off into a separate function does make sense, as it has one clear purpose and because it helps legibility.

wolfreak99 commented:
Okay, I'm not sure how to fix duncans code, it's giving some unhandled exception error with the memmove line.

Gymnasiast commented:
Could you do a rebase? It makes it easier to check and test your code.

marijnvdwerf commented on src/windows/cheats.c:
```diff
@@ -861,9 +1005,12 @@ static void window_cheats_paint(rct_window *w, rct_drawpixelinfo *dpi)
+		if (widget_is_disabled(w, WIDX_MONEY_SPINNER)) {
+			colour |= 0x40;
+		}
+		gfx_draw_string_left(dpi, 1390, gCommonFormatArgs, colour, w->x + XPL(0) + TXTO, w->y + YPL(2) + TXTO);
```
`STR_BOTTOM_TOOLBAR_CASH`

janisozaur commented on src/windows/cheats.c:
```diff
@@ -499,6 +565,81 @@ static void window_cheats_misc_dropdown(rct_window *w, int widgetIndex, int drop
+						money_spinner_value = MONEY(number, decimal);
+
+						free(tokenize);
+						free(numberText);
```
Why are you `free`ing this?

wolfreak99 replied:
I'm not that knowledged on C and i was kinda playing it safe here. should i remove free(numberText) AND free(tokenize) or just free(numberText)?

Gymnasiast replied:
As far as I know, you must free things you allocated via malloc (and related functions). If you didn't use malloc, you mustn't free.

janisozaur replied:
[`strtok` documentation ](http://man7.org/linux/man-pages/man3/strtok.3.html) does not say anything about user having to `free` returned pointer. Moreover, it says (at least for glibc's version):

> The strtok() function uses a static buffer while parsing, so it's not thread safe.

You ought not `free` any static memory.

janisozaur commented on src/windows/cheats.c:
```diff
@@ -499,6 +565,81 @@ static void window_cheats_misc_dropdown(rct_window *w, int widgetIndex, int drop
+	CONVERT_STRING_TO_NUMBER,
+};
+
+static void window_cheats_convert_money(int convertMethod, char* text) 
```
This whole function is incredibly complex for what it does. I'd much rather see a `money_to_string` and `string_to_money` conversion methods, rather than this. They could consume a pointer to whatever string buffer is required.
As for string to money conversion: why can't you just find decimal separator, `atoi` everything you can to the left of it, then to the right of it? Consider using `sscanf` for it, as it returns number of bytes/chars consumed.

wolfreak99 replied:
" They could consume a pointer to whatever string buffer is required." I am not sure what you mean by this, could you clarify? and i'll work on pushing this to "money_to_string" and "string_to_money" like you've mentioned and migrate it to utils.c (or util.c, whichever it's called).

"sscanf for it, as it returns number of bytes/chars consumed." could you also clear this up? if i do your method (about finding a decimal seperator, atoi'ing everything to the left and then right), would i have to make sure everything that's not a number or the decimal seperator gets filtered out beforehand? i would honestly like to make this much simpler because it looks so complex for just a miniscule thing.

janisozaur replied:
`money string_to_money(char * string_to_monetize, size_t len)`
`void money_to_string(money amount, char * buffer_to_put_value_to, size_t buffer_len)`

wolfreak99 commented:
@janisozaur  I have updated the code to reflect your requested changes, with the exception of your suggestion for string_to_money (i asked what you meant by sscanf, and you never answered so i'm still unsure of what you want from me there). could you take off the requested changes thing?
